// ng2-datepicker/__ivy_ngcc__/fesm2015/ng2-datepicker.js
import { InjectionToken, EventEmitter, Component, ElementRef, ChangeDetectorRef, Inject, Input, HostListener, NgModule } from '@angular/core';
import { DOCUMENT, CommonModule } from '@angular/common';
import { NG_VALUE_ACCESSOR, FormsModule } from '@angular/forms';
import { SlimScrollOptions, SlimScrollEvent, NgSlimScrollModule } from 'ngx-slimscroll';
import { getYear, format, addMonths, subMonths, addYears, subYears, setYear, startOfMonth, endOfMonth, eachDayOfInterval, getDay, subDays, setDay, getDate, getMonth, isToday, isSameDay, isSameMonth, isSameYear, isBefore, isAfter } from 'date-fns';
import locale from 'date-fns/locale/en-US';
import { Subscription, fromEvent } from 'rxjs';
import { filter } from 'rxjs/operators';

import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@angular/common';
import * as ɵngcc2 from '@angular/forms';
import * as ɵngcc3 from 'ngx-slimscroll';

function DatepickerComponent_div_2__svg_svg_7_Template(rf, ctx) { if (rf & 1) {
    const _r6 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵnamespaceSVG();
    ɵngcc0.ɵɵelementStart(0, "svg", 13);
    ɵngcc0.ɵɵlistener("click", function DatepickerComponent_div_2__svg_svg_7_Template__svg_svg_click_0_listener() { ɵngcc0.ɵɵrestoreView(_r6); const ctx_r5 = ɵngcc0.ɵɵnextContext(2); return ɵngcc0.ɵɵresetView(ctx_r5.prevMonth()); });
    ɵngcc0.ɵɵelement(1, "path", 14);
    ɵngcc0.ɵɵelementEnd();
} }
function DatepickerComponent_div_2__svg_svg_8_Template(rf, ctx) { if (rf & 1) {
    const _r8 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵnamespaceSVG();
    ɵngcc0.ɵɵelementStart(0, "svg", 15);
    ɵngcc0.ɵɵlistener("click", function DatepickerComponent_div_2__svg_svg_8_Template__svg_svg_click_0_listener() { ɵngcc0.ɵɵrestoreView(_r8); const ctx_r7 = ɵngcc0.ɵɵnextContext(2); return ɵngcc0.ɵɵresetView(ctx_r7.nextMonth()); });
    ɵngcc0.ɵɵelement(1, "path", 16);
    ɵngcc0.ɵɵelementEnd();
} }
function DatepickerComponent_div_2_div_9_span_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "span", 22);
    ɵngcc0.ɵɵtext(1);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const day_r11 = ctx.$implicit;
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate(day_r11);
} }
const _c0 = function (a0, a1, a2, a3) { return { "is-prev-month": a0, "is-today": a1, "is-selected": a2, "is-disabled": a3 }; };
function DatepickerComponent_div_2_div_9_span_4_Template(rf, ctx) { if (rf & 1) {
    const _r15 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "span", 23);
    ɵngcc0.ɵɵlistener("click", function DatepickerComponent_div_2_div_9_span_4_Template_span_click_0_listener() { const restoredCtx = ɵngcc0.ɵɵrestoreView(_r15); const day_r12 = restoredCtx.$implicit; const i_r13 = restoredCtx.index; const ctx_r14 = ɵngcc0.ɵɵnextContext(3); return ɵngcc0.ɵɵresetView(day_r12.isSelectable && ctx_r14.setDate(i_r13)); });
    ɵngcc0.ɵɵtext(1);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const day_r12 = ctx.$implicit;
    ɵngcc0.ɵɵproperty("ngClass", ɵngcc0.ɵɵpureFunction4(2, _c0, !day_r12.inThisMonth, day_r12.isToday, day_r12.isSelected, !day_r12.isSelectable));
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate(day_r12.day);
} }
function DatepickerComponent_div_2_div_9_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 17)(1, "div", 18);
    ɵngcc0.ɵɵtemplate(2, DatepickerComponent_div_2_div_9_span_2_Template, 2, 1, "span", 19);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(3, "div", 20);
    ɵngcc0.ɵɵtemplate(4, DatepickerComponent_div_2_div_9_span_4_Template, 2, 7, "span", 21);
    ɵngcc0.ɵɵelementEnd()();
} if (rf & 2) {
    const ctx_r3 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("ngForOf", ctx_r3.dayNames);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("ngForOf", ctx_r3.days);
} }
const _c1 = function (a0) { return { "is-selected": a0 }; };
function DatepickerComponent_div_2_div_10_span_2_Template(rf, ctx) { if (rf & 1) {
    const _r20 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "span", 27);
    ɵngcc0.ɵɵlistener("click", function DatepickerComponent_div_2_div_10_span_2_Template_span_click_0_listener() { const restoredCtx = ɵngcc0.ɵɵrestoreView(_r20); const i_r18 = restoredCtx.index; const ctx_r19 = ɵngcc0.ɵɵnextContext(3); return ɵngcc0.ɵɵresetView(ctx_r19.setYear(i_r18)); });
    ɵngcc0.ɵɵtext(1);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const year_r17 = ctx.$implicit;
    ɵngcc0.ɵɵproperty("ngClass", ɵngcc0.ɵɵpureFunction1(2, _c1, year_r17.isThisYear));
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate(year_r17.year);
} }
function DatepickerComponent_div_2_div_10_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 24)(1, "div", 25);
    ɵngcc0.ɵɵtemplate(2, DatepickerComponent_div_2_div_10_span_2_Template, 2, 4, "span", 26);
    ɵngcc0.ɵɵelementEnd()();
} if (rf & 2) {
    const ctx_r4 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("options", ctx_r4.scrollOptions)("scrollEvents", ctx_r4.scrollEvents);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngForOf", ctx_r4.years);
} }
function DatepickerComponent_div_2_Template(rf, ctx) { if (rf & 1) {
    const _r22 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "div", 3)(1, "div", 4)(2, "div", 5)(3, "span", 6)(4, "span", 7);
    ɵngcc0.ɵɵlistener("click", function DatepickerComponent_div_2_Template_span_click_4_listener() { ɵngcc0.ɵɵrestoreView(_r22); const ctx_r21 = ɵngcc0.ɵɵnextContext(); return ɵngcc0.ɵɵresetView(ctx_r21.toggleView()); });
    ɵngcc0.ɵɵtext(5);
    ɵngcc0.ɵɵelementEnd()()();
    ɵngcc0.ɵɵelementStart(6, "div", 8);
    ɵngcc0.ɵɵtemplate(7, DatepickerComponent_div_2__svg_svg_7_Template, 2, 0, "svg", 9);
    ɵngcc0.ɵɵtemplate(8, DatepickerComponent_div_2__svg_svg_8_Template, 2, 0, "svg", 10);
    ɵngcc0.ɵɵelementEnd()();
    ɵngcc0.ɵɵtemplate(9, DatepickerComponent_div_2_div_9_Template, 5, 2, "div", 11);
    ɵngcc0.ɵɵtemplate(10, DatepickerComponent_div_2_div_10_Template, 3, 3, "div", 12);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r0 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(5);
    ɵngcc0.ɵɵtextInterpolate(ctx_r0.title);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r0.view === "days");
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r0.view === "days");
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r0.view === "days");
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r0.view === "years");
} }
const DATEPICKER_OPTIONS = new InjectionToken('Datepicker config');
function mergeDatepickerOptions(opts) {
    return Object.assign(Object.assign({}, defaultOptions), opts);
}
const defaultOptions = {
    minDate: null,
    maxDate: null,
    minYear: getYear(new Date()) - 30,
    maxYear: getYear(new Date()) + 30,
    placeholder: '',
    format: 'LLLL do yyyy',
    formatTitle: 'LLLL yyyy',
    formatDays: 'EEEEE',
    firstCalendarDay: 0,
    locale,
    position: 'bottom',
    inputClass: '',
    calendarClass: 'datepicker-default',
    scrollBarColor: '#dfe3e9',
    enableKeyboard: true
};

class DatepickerComponent {
    constructor(elementRef, ref, document) {
        this.elementRef = elementRef;
        this.ref = ref;
        this.options = Object.assign({}, defaultOptions);
        this.scrollOptions = new SlimScrollOptions(this.scrollBarOptions);
        this.isOpened = false;
        this.innerValue = new Date();
        this.displayValue = '';
        this.view = 'days';
        this.date = new Date();
        this.years = [];
        this.days = [];
        this.dayNames = [];
        this.scrollEvents = new EventEmitter();
        this.sub = new Subscription();
        this.onTouchedCallback = () => { };
        this.onChangeCallback = () => { };
        this.doc = document;
    }
    get value() {
        return this.innerValue;
    }
    set value(val) {
        this.innerValue = val;
        this.displayValue = format(this.innerValue, this.options.format, { locale: this.options.locale });
        this.onChangeCallback(this.innerValue);
    }
    get title() {
        return format(this.date, this.options.formatTitle);
    }
    get scrollBarOptions() {
        return {
            barBackground: (this.options && this.options.scrollBarColor) || '#dfe3e9',
            gridBackground: 'transparent',
            barBorderRadius: '3',
            gridBorderRadius: '3',
            barWidth: '6',
            gridWidth: '6',
            barMargin: '0',
            gridMargin: '0'
        };
    }
    ngOnInit() {
        this.view = 'days';
        this.date = new Date();
        this.init();
    }
    ngOnChanges(changes) {
        if ('options' in changes) {
            this.options = mergeDatepickerOptions(this.options);
            this.scrollOptions = new SlimScrollOptions(this.scrollBarOptions);
            if (this.sub) {
                this.sub.unsubscribe();
            }
            if (this.options.enableKeyboard) {
                this.sub = fromEvent(this.doc || document, 'keyup')
                    .pipe(filter(() => this.isOpened))
                    .subscribe(e => {
                    e.preventDefault();
                    e.stopPropagation();
                    switch (e.key) {
                        case 'Down':
                        case 'ArrowDown':
                            this.prevYear();
                            break;
                        case 'Up':
                        case 'ArrowUp':
                            this.nextYear();
                            break;
                        case 'Left':
                        case 'ArrowLeft':
                            this.prevMonth();
                            break;
                        case 'Right':
                        case 'ArrowRight':
                            this.nextMonth();
                            break;
                        case 'Esc':
                        case 'Escape':
                        case 'Enter':
                            this.isOpened = false;
                            break;
                        default:
                            return;
                    }
                });
            }
        }
    }
    ngOnDestroy() {
        this.sub.unsubscribe();
    }
    toggle() {
        this.isOpened = !this.isOpened;
        if (this.isOpened) {
            this.view = 'days';
            this.date = this.value;
            this.initDays();
        }
    }
    toggleView() {
        this.view = this.view === 'days' ? 'years' : 'days';
        if (this.view === 'years') {
            this.ref.detectChanges();
            this.scrollToYear();
        }
    }
    nextMonth() {
        this.date = addMonths(this.date, 1);
        this.initDays();
    }
    prevMonth() {
        this.date = subMonths(this.date, 1);
        this.initDays();
    }
    nextYear() {
        this.date = addYears(this.date, 1);
        this.initDays();
    }
    prevYear() {
        this.date = subYears(this.date, 1);
        this.initDays();
    }
    setDate(i) {
        this.date = this.days[i].date;
        this.value = this.date;
        this.initDays();
        this.isOpened = false;
    }
    setYear(i) {
        this.date = setYear(this.date, this.years[i].year);
        this.initDays();
        this.initYears();
        this.view = 'days';
    }
    scrollToYear() {
        const parent = this.elementRef.nativeElement.querySelector('.main-calendar-years');
        const el = this.elementRef.nativeElement.querySelector('.year-unit.is-selected');
        const y = el.offsetTop - parent.clientHeight / 2 + el.clientHeight / 2;
        const event = new SlimScrollEvent({ type: 'scrollTo', y, duration: 100 });
        this.scrollEvents.emit(event);
    }
    init() {
        this.initDayNames();
        this.initDays();
        this.initYears();
    }
    initDays() {
        const date = this.date || new Date();
        const [start, end] = [startOfMonth(date), endOfMonth(date)];
        this.days = eachDayOfInterval({ start, end }).map((d) => this.generateDay(d));
        const tmp = getDay(start) - this.options.firstCalendarDay;
        const prevDays = tmp < 0 ? 7 - this.options.firstCalendarDay : tmp;
        for (let i = 1; i <= prevDays; i++) {
            const d = subDays(start, i);
            this.days.unshift(this.generateDay(d, false));
        }
    }
    initYears() {
        const range = this.options.maxYear - this.options.minYear + 1;
        this.years = Array.from(new Array(range), (_, i) => i + this.options.minYear).map(year => {
            return { year, isThisYear: year === getYear(this.date) };
        });
    }
    initDayNames() {
        this.dayNames = [];
        const start = this.options.firstCalendarDay;
        for (let i = start; i <= 6 + start; i++) {
            const date = setDay(new Date(), i);
            this.dayNames.push(format(date, this.options.formatDays, { locale: this.options.locale }));
        }
    }
    generateDay(date, inThisMonth = true) {
        return {
            date,
            day: getDate(date),
            month: getMonth(date),
            year: getYear(date),
            inThisMonth,
            isToday: isToday(date),
            isSelected: isSameDay(date, this.innerValue) && isSameMonth(date, this.innerValue) && isSameYear(date, this.innerValue),
            isSelectable: this.isDateSelectable(date)
        };
    }
    isDateSelectable(date) {
        if (this.options.minDate && isBefore(date, this.options.minDate)) {
            return false;
        }
        if (this.options.maxDate && isAfter(date, this.options.maxDate)) {
            return false;
        }
        return true;
    }
    writeValue(val) {
        if (!val) {
            return;
        }
        this.innerValue = val;
        this.displayValue = format(this.innerValue, this.options.format, { locale: this.options.locale });
        this.init();
    }
    registerOnChange(fn) {
        this.onChangeCallback = fn;
    }
    registerOnTouched(fn) {
        this.onTouchedCallback = fn;
    }
    onBlur(e) {
        if (!this.isOpened) {
            return;
        }
        const input = this.elementRef.nativeElement.querySelector('.datepicker-container > input');
        if (!input || e.target === input || input.contains(e.target)) {
            return;
        }
        const container = this.elementRef.nativeElement.querySelector('.datepicker-container > .calendar-container');
        if (container &&
            container !== e.target &&
            !container.contains(e.target) &&
            !e.target.classList.contains('year-unit')) {
            this.isOpened = false;
        }
    }
}
DatepickerComponent.ɵfac = function DatepickerComponent_Factory(t) { return new (t || DatepickerComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ElementRef), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ChangeDetectorRef), ɵngcc0.ɵɵdirectiveInject(DOCUMENT)); };
DatepickerComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: DatepickerComponent, selectors: [["ngx-datepicker"]], hostBindings: function DatepickerComponent_HostBindings(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵlistener("click", function DatepickerComponent_click_HostBindingHandler($event) { return ctx.onBlur($event); }, false, ɵngcc0.ɵɵresolveDocument);
    } }, inputs: { options: "options", scrollOptions: "scrollOptions", isOpened: "isOpened" }, features: [ɵngcc0.ɵɵProvidersFeature([{ provide: NG_VALUE_ACCESSOR, useExisting: DatepickerComponent, multi: true }]), ɵngcc0.ɵɵNgOnChangesFeature], decls: 3, vars: 7, consts: [[1, "datepicker-container"], ["type", "text", "readonly", "", 3, "ngModel", "placeholder", "ngModelChange", "click"], ["class", "calendar-container", 4, "ngIf"], [1, "calendar-container"], [1, "top-container"], [1, "month-year-container"], [1, "month-year-text"], [3, "click"], [1, "controls"], ["xmlns", "http://www.w3.org/2000/svg", "class", "control prev-month", "width", "8", "height", "13", 3, "click", 4, "ngIf"], ["xmlns", "http://www.w3.org/2000/svg", "class", "control next-month", "width", "8", "height", "13", 3, "click", 4, "ngIf"], ["class", "main-calendar-container is-days", 4, "ngIf"], ["class", "main-calendar-container is-years", 4, "ngIf"], ["xmlns", "http://www.w3.org/2000/svg", "width", "8", "height", "13", 1, "control", "prev-month", 3, "click"], ["d", "M7.575 1.131a1.063 1.063 0 00-1.502 0l-4.93 4.93c-.42.42-.42 1.099 0 1.518l4.93 4.93a1.063 1.063 0 001.503-1.503L3.388 6.82l4.186-4.186a1.063 1.063 0 000-1.503z"], ["xmlns", "http://www.w3.org/2000/svg", "width", "8", "height", "13", 1, "control", "next-month", 3, "click"], ["d", "M1.14 1.131a1.063 1.063 0 011.502 0l4.93 4.93c.42.42.42 1.099 0 1.518l-4.93 4.93a1.063 1.063 0 01-1.503-1.503L5.327 6.82 1.14 2.634a1.063 1.063 0 010-1.503z"], [1, "main-calendar-container", "is-days"], [1, "main-calendar-day-names"], ["class", "day-name-unit", 4, "ngFor", "ngForOf"], [1, "main-calendar-days"], ["class", "day-unit", 3, "ngClass", "click", 4, "ngFor", "ngForOf"], [1, "day-name-unit"], [1, "day-unit", 3, "ngClass", "click"], [1, "main-calendar-container", "is-years"], ["slimScroll", "", 1, "main-calendar-years", 3, "options", "scrollEvents"], ["class", "year-unit", 3, "ngClass", "click", 4, "ngFor", "ngForOf"], [1, "year-unit", 3, "ngClass", "click"]], template: function DatepickerComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "div", 0)(1, "input", 1);
        ɵngcc0.ɵɵlistener("ngModelChange", function DatepickerComponent_Template_input_ngModelChange_1_listener($event) { return ctx.displayValue = $event; })("click", function DatepickerComponent_Template_input_click_1_listener() { return ctx.toggle(); });
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵtemplate(2, DatepickerComponent_div_2_Template, 11, 5, "div", 2);
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵclassMap(ctx.options.calendarClass);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵclassMap(ctx.options.inputClass);
        ɵngcc0.ɵɵproperty("ngModel", ctx.displayValue)("placeholder", ctx.options.placeholder);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.isOpened);
    } }, dependencies: [ɵngcc1.NgClass, ɵngcc1.NgForOf, ɵngcc1.NgIf, ɵngcc2.DefaultValueAccessor, ɵngcc2.NgControlStatus, ɵngcc2.NgModel, ɵngcc3.SlimScrollDirective], styles: [".datepicker-container[_ngcontent-%COMP%]{position:relative;display:inline-block;box-sizing:border-box}.datepicker-container[_ngcontent-%COMP%]   *[_ngcontent-%COMP%]{box-sizing:inherit}.datepicker-container[_ngcontent-%COMP%]   input[_ngcontent-%COMP%]{-webkit-touch-callout:none;-webkit-user-select:none;-moz-user-select:none;-ms-user-select:none;user-select:none;cursor:pointer}.datepicker-container[_ngcontent-%COMP%]   .is-hidden[_ngcontent-%COMP%]{display:none!important}.datepicker-container[_ngcontent-%COMP%]   .calendar-container[_ngcontent-%COMP%]{position:absolute;width:320px;border-radius:4px;top:35px;left:0;z-index:10}.datepicker-container[_ngcontent-%COMP%]   .calendar-container[_ngcontent-%COMP%]   .top-container[_ngcontent-%COMP%]{-webkit-touch-callout:none;-webkit-user-select:none;-moz-user-select:none;-ms-user-select:none;user-select:none;width:100%;padding:20px;display:flex;justify-content:space-between;align-items:center}.datepicker-container[_ngcontent-%COMP%]   .calendar-container[_ngcontent-%COMP%]   .top-container[_ngcontent-%COMP%]   .month-year-container[_ngcontent-%COMP%]{height:100%;display:flex;align-items:center;cursor:pointer;position:relative}.datepicker-container[_ngcontent-%COMP%]   .calendar-container[_ngcontent-%COMP%]   .top-container[_ngcontent-%COMP%]   .month-year-container[_ngcontent-%COMP%]   .month-year-text[_ngcontent-%COMP%]{font-size:16px;width:100%}.datepicker-container[_ngcontent-%COMP%]   .calendar-container[_ngcontent-%COMP%]   .top-container[_ngcontent-%COMP%]   .controls[_ngcontent-%COMP%]{height:100%;display:flex;justify-content:flex-end;align-items:center}.datepicker-container[_ngcontent-%COMP%]   .calendar-container[_ngcontent-%COMP%]   .top-container[_ngcontent-%COMP%]   .controls[_ngcontent-%COMP%]   .control[_ngcontent-%COMP%]{cursor:pointer;margin-left:20px}.datepicker-container[_ngcontent-%COMP%]   .calendar-container[_ngcontent-%COMP%]   .main-calendar-container[_ngcontent-%COMP%]{width:100%;height:100%;font-size:12px;display:block}.datepicker-container[_ngcontent-%COMP%]   .calendar-container[_ngcontent-%COMP%]   .main-calendar-container[_ngcontent-%COMP%]   .main-calendar-day-names[_ngcontent-%COMP%]{-webkit-touch-callout:none;-webkit-user-select:none;-moz-user-select:none;-ms-user-select:none;user-select:none;display:flex;align-items:center;height:35px;padding:0 10px}.datepicker-container[_ngcontent-%COMP%]   .calendar-container[_ngcontent-%COMP%]   .main-calendar-container[_ngcontent-%COMP%]   .main-calendar-day-names[_ngcontent-%COMP%]   .day-name-unit[_ngcontent-%COMP%]{width:calc(100% / 7);text-transform:uppercase;text-align:center}.datepicker-container[_ngcontent-%COMP%]   .calendar-container[_ngcontent-%COMP%]   .main-calendar-container[_ngcontent-%COMP%]   .main-calendar-days[_ngcontent-%COMP%], .datepicker-container[_ngcontent-%COMP%]   .calendar-container[_ngcontent-%COMP%]   .main-calendar-container[_ngcontent-%COMP%]   .main-calendar-years[_ngcontent-%COMP%]{-webkit-touch-callout:none;-webkit-user-select:none;-moz-user-select:none;-ms-user-select:none;user-select:none;padding:10px;width:100%;overflow:hidden;text-align:left;display:inline-block}.datepicker-container[_ngcontent-%COMP%]   .calendar-container[_ngcontent-%COMP%]   .main-calendar-container[_ngcontent-%COMP%]   .main-calendar-days.is-hidden[_ngcontent-%COMP%], .datepicker-container[_ngcontent-%COMP%]   .calendar-container[_ngcontent-%COMP%]   .main-calendar-container[_ngcontent-%COMP%]   .main-calendar-years.is-hidden[_ngcontent-%COMP%]{display:none}.datepicker-container[_ngcontent-%COMP%]   .calendar-container[_ngcontent-%COMP%]   .main-calendar-container[_ngcontent-%COMP%]   .main-calendar-days[_ngcontent-%COMP%]   .day-unit[_ngcontent-%COMP%], .datepicker-container[_ngcontent-%COMP%]   .calendar-container[_ngcontent-%COMP%]   .main-calendar-container[_ngcontent-%COMP%]   .main-calendar-days[_ngcontent-%COMP%]   .year-unit[_ngcontent-%COMP%], .datepicker-container[_ngcontent-%COMP%]   .calendar-container[_ngcontent-%COMP%]   .main-calendar-container[_ngcontent-%COMP%]   .main-calendar-years[_ngcontent-%COMP%]   .day-unit[_ngcontent-%COMP%], .datepicker-container[_ngcontent-%COMP%]   .calendar-container[_ngcontent-%COMP%]   .main-calendar-container[_ngcontent-%COMP%]   .main-calendar-years[_ngcontent-%COMP%]   .year-unit[_ngcontent-%COMP%]{width:calc(100% / 7);display:inline-flex;align-items:center;justify-content:center;cursor:pointer;-webkit-user-select:none;-moz-user-select:none;-ms-user-select:none;user-select:none;font-size:13px;border-radius:50%}.datepicker-container[_ngcontent-%COMP%]   .calendar-container[_ngcontent-%COMP%]   .main-calendar-container[_ngcontent-%COMP%]   .main-calendar-days[_ngcontent-%COMP%]   .day-unit[_ngcontent-%COMP%]:before, .datepicker-container[_ngcontent-%COMP%]   .calendar-container[_ngcontent-%COMP%]   .main-calendar-container[_ngcontent-%COMP%]   .main-calendar-days[_ngcontent-%COMP%]   .year-unit[_ngcontent-%COMP%]:before, .datepicker-container[_ngcontent-%COMP%]   .calendar-container[_ngcontent-%COMP%]   .main-calendar-container[_ngcontent-%COMP%]   .main-calendar-years[_ngcontent-%COMP%]   .day-unit[_ngcontent-%COMP%]:before, .datepicker-container[_ngcontent-%COMP%]   .calendar-container[_ngcontent-%COMP%]   .main-calendar-container[_ngcontent-%COMP%]   .main-calendar-years[_ngcontent-%COMP%]   .year-unit[_ngcontent-%COMP%]:before{content:\"\";float:left;padding-top:calc(100% - 5px)}.datepicker-container[_ngcontent-%COMP%]   .calendar-container[_ngcontent-%COMP%]   .main-calendar-container[_ngcontent-%COMP%]   .main-calendar-days[_ngcontent-%COMP%]   .day-unit.is-disabled[_ngcontent-%COMP%], .datepicker-container[_ngcontent-%COMP%]   .calendar-container[_ngcontent-%COMP%]   .main-calendar-container[_ngcontent-%COMP%]   .main-calendar-days[_ngcontent-%COMP%]   .year-unit.is-disabled[_ngcontent-%COMP%], .datepicker-container[_ngcontent-%COMP%]   .calendar-container[_ngcontent-%COMP%]   .main-calendar-container[_ngcontent-%COMP%]   .main-calendar-years[_ngcontent-%COMP%]   .day-unit.is-disabled[_ngcontent-%COMP%], .datepicker-container[_ngcontent-%COMP%]   .calendar-container[_ngcontent-%COMP%]   .main-calendar-container[_ngcontent-%COMP%]   .main-calendar-years[_ngcontent-%COMP%]   .year-unit.is-disabled[_ngcontent-%COMP%]{cursor:not-allowed}.datepicker-container[_ngcontent-%COMP%]   .calendar-container[_ngcontent-%COMP%]   .main-calendar-container[_ngcontent-%COMP%]   .main-calendar-days[_ngcontent-%COMP%]   .day-unit.is-disabled[_ngcontent-%COMP%]:hover, .datepicker-container[_ngcontent-%COMP%]   .calendar-container[_ngcontent-%COMP%]   .main-calendar-container[_ngcontent-%COMP%]   .main-calendar-days[_ngcontent-%COMP%]   .year-unit.is-disabled[_ngcontent-%COMP%]:hover, .datepicker-container[_ngcontent-%COMP%]   .calendar-container[_ngcontent-%COMP%]   .main-calendar-container[_ngcontent-%COMP%]   .main-calendar-years[_ngcontent-%COMP%]   .day-unit.is-disabled[_ngcontent-%COMP%]:hover, .datepicker-container[_ngcontent-%COMP%]   .calendar-container[_ngcontent-%COMP%]   .main-calendar-container[_ngcontent-%COMP%]   .main-calendar-years[_ngcontent-%COMP%]   .year-unit.is-disabled[_ngcontent-%COMP%]:hover{background:transparent}.datepicker-container[_ngcontent-%COMP%]   .calendar-container[_ngcontent-%COMP%]   .main-calendar-container[_ngcontent-%COMP%]   .main-calendar-years[_ngcontent-%COMP%]{display:block;padding:10px;height:275px}.datepicker-container[_ngcontent-%COMP%]   .calendar-container[_ngcontent-%COMP%]   .main-calendar-container[_ngcontent-%COMP%]   .main-calendar-years[_ngcontent-%COMP%]   .year-unit[_ngcontent-%COMP%]{width:calc(100% / 4);height:35px;border-radius:20px;margin:8px 0}.datepicker-container[_ngcontent-%COMP%]   .calendar-container[_ngcontent-%COMP%]   .main-calendar-container[_ngcontent-%COMP%]   .main-calendar-years[_ngcontent-%COMP%]   .year-unit[_ngcontent-%COMP%]:before{padding-top:0}.datepicker-default[_ngcontent-%COMP%]   .calendar-container[_ngcontent-%COMP%]{background:#fff;border:1px solid #eaedf3;box-shadow:0 4px 12px rgba(0,0,0,.05);top:35px;left:0;font-weight:600}.datepicker-default[_ngcontent-%COMP%]   .month-year-text[_ngcontent-%COMP%]{color:#010001}.datepicker-default[_ngcontent-%COMP%]   .control[_ngcontent-%COMP%]   path[_ngcontent-%COMP%]{fill:#aaa8ab}.datepicker-default[_ngcontent-%COMP%]   .control[_ngcontent-%COMP%]:hover   path[_ngcontent-%COMP%]{fill:#010001}.datepicker-default[_ngcontent-%COMP%]   .main-calendar-day-names[_ngcontent-%COMP%]{border-bottom:1px solid #eaedf3}.datepicker-default[_ngcontent-%COMP%]   .day-name-unit[_ngcontent-%COMP%]{color:#aaa8ab}.datepicker-default[_ngcontent-%COMP%]   .day-unit[_ngcontent-%COMP%], .datepicker-default[_ngcontent-%COMP%]   .year-unit[_ngcontent-%COMP%]{color:#010001}.datepicker-default[_ngcontent-%COMP%]   .day-unit.is-prev-month[_ngcontent-%COMP%], .datepicker-default[_ngcontent-%COMP%]   .year-unit.is-prev-month[_ngcontent-%COMP%]{color:#bbbabe}.datepicker-default[_ngcontent-%COMP%]   .day-unit.is-today[_ngcontent-%COMP%], .datepicker-default[_ngcontent-%COMP%]   .year-unit.is-today[_ngcontent-%COMP%]{background:#edeef2}.datepicker-default[_ngcontent-%COMP%]   .day-unit.is-selected[_ngcontent-%COMP%], .datepicker-default[_ngcontent-%COMP%]   .day-unit[_ngcontent-%COMP%]:hover, .datepicker-default[_ngcontent-%COMP%]   .year-unit.is-selected[_ngcontent-%COMP%], .datepicker-default[_ngcontent-%COMP%]   .year-unit[_ngcontent-%COMP%]:hover{background:#010001;color:#fff}.datepicker-default[_ngcontent-%COMP%]   .day-unit.is-disabled[_ngcontent-%COMP%], .datepicker-default[_ngcontent-%COMP%]   .year-unit.is-disabled[_ngcontent-%COMP%]{color:#aaa8ab}.datepicker-default[_ngcontent-%COMP%]   .day-unit.is-disabled[_ngcontent-%COMP%]:hover, .datepicker-default[_ngcontent-%COMP%]   .year-unit.is-disabled[_ngcontent-%COMP%]:hover{background:transparent}"] });
DatepickerComponent.ctorParameters = () => [
    { type: ElementRef },
    { type: ChangeDetectorRef },
    { type: undefined, decorators: [{ type: Inject, args: [DOCUMENT,] }] }
];
DatepickerComponent.propDecorators = {
    options: [{ type: Input }],
    scrollOptions: [{ type: Input }],
    isOpened: [{ type: Input }],
    onBlur: [{ type: HostListener, args: ['document:click', ['$event'],] }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(DatepickerComponent, [{
        type: Component,
        args: [{ selector: 'ngx-datepicker', template: "<div class=\"datepicker-container\" [class]=\"options.calendarClass\">\n  <input\n    type=\"text\"\n    [(ngModel)]=\"displayValue\"\n    [class]=\"options.inputClass\"\n    [placeholder]=\"options.placeholder\"\n    readonly\n    (click)=\"toggle()\"\n  />\n\n  <div class=\"calendar-container\" *ngIf=\"isOpened\">\n    <div class=\"top-container\">\n      <div class=\"month-year-container\">\n        <span class=\"month-year-text\">\n          <span (click)=\"toggleView()\">{{ title }}</span>\n        </span>\n      </div>\n      <div class=\"controls\">\n        <svg\n          xmlns=\"http://www.w3.org/2000/svg\"\n          class=\"control prev-month\"\n          width=\"8\"\n          height=\"13\"\n          (click)=\"prevMonth()\"\n          *ngIf=\"view === 'days'\"\n        >\n          <path\n            d=\"M7.575 1.131a1.063 1.063 0 00-1.502 0l-4.93 4.93c-.42.42-.42 1.099 0 1.518l4.93 4.93a1.063 1.063 0 001.503-1.503L3.388 6.82l4.186-4.186a1.063 1.063 0 000-1.503z\"\n          />\n        </svg>\n        <svg\n          xmlns=\"http://www.w3.org/2000/svg\"\n          class=\"control next-month\"\n          width=\"8\"\n          height=\"13\"\n          (click)=\"nextMonth()\"\n          *ngIf=\"view === 'days'\"\n        >\n          <path\n            d=\"M1.14 1.131a1.063 1.063 0 011.502 0l4.93 4.93c.42.42.42 1.099 0 1.518l-4.93 4.93a1.063 1.063 0 01-1.503-1.503L5.327 6.82 1.14 2.634a1.063 1.063 0 010-1.503z\"\n          />\n        </svg>\n      </div>\n    </div>\n    <div class=\"main-calendar-container is-days\" *ngIf=\"view === 'days'\">\n      <div class=\"main-calendar-day-names\">\n        <span class=\"day-name-unit\" *ngFor=\"let day of dayNames\">{{ day }}</span>\n      </div>\n      <div class=\"main-calendar-days\">\n        <span\n          class=\"day-unit\"\n          *ngFor=\"let day of days; let i = index\"\n          [ngClass]=\"{\n            'is-prev-month': !day.inThisMonth,\n            'is-today': day.isToday,\n            'is-selected': day.isSelected,\n            'is-disabled': !day.isSelectable\n          }\"\n          (click)=\"day.isSelectable && setDate(i)\"\n          >{{ day.day }}</span\n        >\n      </div>\n    </div>\n    <div class=\"main-calendar-container is-years\" *ngIf=\"view === 'years'\">\n      <div class=\"main-calendar-years\" slimScroll [options]=\"scrollOptions\" [scrollEvents]=\"scrollEvents\">\n        <span\n          class=\"year-unit\"\n          *ngFor=\"let year of years; let i = index\"\n          [ngClass]=\"{ 'is-selected': year.isThisYear }\"\n          (click)=\"setYear(i)\"\n          >{{ year.year }}</span\n        >\n      </div>\n    </div>\n  </div>\n</div>\n", providers: [{ provide: NG_VALUE_ACCESSOR, useExisting: DatepickerComponent, multi: true }], styles: [".datepicker-container{position:relative;display:inline-block;box-sizing:border-box}.datepicker-container *{box-sizing:inherit}.datepicker-container input{-webkit-touch-callout:none;-webkit-user-select:none;-moz-user-select:none;-ms-user-select:none;user-select:none;cursor:pointer}.datepicker-container .is-hidden{display:none!important}.datepicker-container .calendar-container{position:absolute;width:320px;border-radius:4px;top:35px;left:0;z-index:10}.datepicker-container .calendar-container .top-container{-webkit-touch-callout:none;-webkit-user-select:none;-moz-user-select:none;-ms-user-select:none;user-select:none;width:100%;padding:20px;display:flex;justify-content:space-between;align-items:center}.datepicker-container .calendar-container .top-container .month-year-container{height:100%;display:flex;align-items:center;cursor:pointer;position:relative}.datepicker-container .calendar-container .top-container .month-year-container .month-year-text{font-size:16px;width:100%}.datepicker-container .calendar-container .top-container .controls{height:100%;display:flex;justify-content:flex-end;align-items:center}.datepicker-container .calendar-container .top-container .controls .control{cursor:pointer;margin-left:20px}.datepicker-container .calendar-container .main-calendar-container{width:100%;height:100%;font-size:12px;display:block}.datepicker-container .calendar-container .main-calendar-container .main-calendar-day-names{-webkit-touch-callout:none;-webkit-user-select:none;-moz-user-select:none;-ms-user-select:none;user-select:none;display:flex;align-items:center;height:35px;padding:0 10px}.datepicker-container .calendar-container .main-calendar-container .main-calendar-day-names .day-name-unit{width:calc(100% / 7);text-transform:uppercase;text-align:center}.datepicker-container .calendar-container .main-calendar-container .main-calendar-days,.datepicker-container .calendar-container .main-calendar-container .main-calendar-years{-webkit-touch-callout:none;-webkit-user-select:none;-moz-user-select:none;-ms-user-select:none;user-select:none;padding:10px;width:100%;overflow:hidden;text-align:left;display:inline-block}.datepicker-container .calendar-container .main-calendar-container .main-calendar-days.is-hidden,.datepicker-container .calendar-container .main-calendar-container .main-calendar-years.is-hidden{display:none}.datepicker-container .calendar-container .main-calendar-container .main-calendar-days .day-unit,.datepicker-container .calendar-container .main-calendar-container .main-calendar-days .year-unit,.datepicker-container .calendar-container .main-calendar-container .main-calendar-years .day-unit,.datepicker-container .calendar-container .main-calendar-container .main-calendar-years .year-unit{width:calc(100% / 7);display:inline-flex;align-items:center;justify-content:center;cursor:pointer;-webkit-user-select:none;-moz-user-select:none;-ms-user-select:none;user-select:none;font-size:13px;border-radius:50%}.datepicker-container .calendar-container .main-calendar-container .main-calendar-days .day-unit:before,.datepicker-container .calendar-container .main-calendar-container .main-calendar-days .year-unit:before,.datepicker-container .calendar-container .main-calendar-container .main-calendar-years .day-unit:before,.datepicker-container .calendar-container .main-calendar-container .main-calendar-years .year-unit:before{content:\"\";float:left;padding-top:calc(100% - 5px)}.datepicker-container .calendar-container .main-calendar-container .main-calendar-days .day-unit.is-disabled,.datepicker-container .calendar-container .main-calendar-container .main-calendar-days .year-unit.is-disabled,.datepicker-container .calendar-container .main-calendar-container .main-calendar-years .day-unit.is-disabled,.datepicker-container .calendar-container .main-calendar-container .main-calendar-years .year-unit.is-disabled{cursor:not-allowed}.datepicker-container .calendar-container .main-calendar-container .main-calendar-days .day-unit.is-disabled:hover,.datepicker-container .calendar-container .main-calendar-container .main-calendar-days .year-unit.is-disabled:hover,.datepicker-container .calendar-container .main-calendar-container .main-calendar-years .day-unit.is-disabled:hover,.datepicker-container .calendar-container .main-calendar-container .main-calendar-years .year-unit.is-disabled:hover{background:transparent}.datepicker-container .calendar-container .main-calendar-container .main-calendar-years{display:block;padding:10px;height:275px}.datepicker-container .calendar-container .main-calendar-container .main-calendar-years .year-unit{width:calc(100% / 4);height:35px;border-radius:20px;margin:8px 0}.datepicker-container .calendar-container .main-calendar-container .main-calendar-years .year-unit:before{padding-top:0}.datepicker-default .calendar-container{background:#fff;border:1px solid #eaedf3;box-shadow:0 4px 12px rgba(0,0,0,.05);top:35px;left:0;font-weight:600}.datepicker-default .month-year-text{color:#010001}.datepicker-default .control path{fill:#aaa8ab}.datepicker-default .control:hover path{fill:#010001}.datepicker-default .main-calendar-day-names{border-bottom:1px solid #eaedf3}.datepicker-default .day-name-unit{color:#aaa8ab}.datepicker-default .day-unit,.datepicker-default .year-unit{color:#010001}.datepicker-default .day-unit.is-prev-month,.datepicker-default .year-unit.is-prev-month{color:#bbbabe}.datepicker-default .day-unit.is-today,.datepicker-default .year-unit.is-today{background:#edeef2}.datepicker-default .day-unit.is-selected,.datepicker-default .day-unit:hover,.datepicker-default .year-unit.is-selected,.datepicker-default .year-unit:hover{background:#010001;color:#fff}.datepicker-default .day-unit.is-disabled,.datepicker-default .year-unit.is-disabled{color:#aaa8ab}.datepicker-default .day-unit.is-disabled:hover,.datepicker-default .year-unit.is-disabled:hover{background:transparent}"] }]
    }], function () { return [{ type: ɵngcc0.ElementRef }, { type: ɵngcc0.ChangeDetectorRef }, { type: undefined, decorators: [{
                type: Inject,
                args: [DOCUMENT]
            }] }]; }, { options: [{
            type: Input
        }], scrollOptions: [{
            type: Input
        }], isOpened: [{
            type: Input
        }], onBlur: [{
            type: HostListener,
            args: ['document:click', ['$event']]
        }] }); })();

class DatepickerModule {
    static forRoot(options) {
        return {
            ngModule: DatepickerModule,
            providers: [{ provide: DATEPICKER_OPTIONS, useValue: mergeDatepickerOptions(options) }]
        };
    }
}
DatepickerModule.ɵfac = function DatepickerModule_Factory(t) { return new (t || DatepickerModule)(); };
DatepickerModule.ɵmod = /*@__PURE__*/ ɵngcc0.ɵɵdefineNgModule({ type: DatepickerModule });
DatepickerModule.ɵinj = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjector({ imports: [CommonModule, FormsModule, NgSlimScrollModule, CommonModule, FormsModule] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(DatepickerModule, [{
        type: NgModule,
        args: [{
                imports: [CommonModule, FormsModule, NgSlimScrollModule],
                declarations: [DatepickerComponent],
                exports: [DatepickerComponent, CommonModule, FormsModule]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(DatepickerModule, { declarations: function () { return [DatepickerComponent]; }, imports: function () { return [CommonModule, FormsModule, NgSlimScrollModule]; }, exports: function () { return [DatepickerComponent, CommonModule, FormsModule]; } }); })();

/**
 * Generated bundle index. Do not edit.
 */

export { DATEPICKER_OPTIONS, DatepickerComponent, DatepickerModule, defaultOptions, mergeDatepickerOptions };
