// ng2-google-charts/esm2015/lib/ng2-google-charts.module.js
import { NgModule } from '@angular/core';
import { GoogleChartComponent } from './google-chart/google-chart.component';
import { GoogleChartsLoaderService } from './google-charts-loader.service';
import { GoogleChartsDashboardComponent } from './google-charts-dashboard/google-charts-dashboard.component';
import { GoogleChartsControlComponent } from './google-charts-control/google-charts-control.component';
import * as i0 from "@angular/core";
export class Ng2GoogleChartsModule {
}
Ng2GoogleChartsModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.14", ngImport: i0, type: Ng2GoogleChartsModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
Ng2GoogleChartsModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.14", ngImport: i0, type: Ng2GoogleChartsModule, declarations: [GoogleChartComponent,
        GoogleChartsDashboardComponent,
        GoogleChartsControlComponent], exports: [GoogleChartComponent,
        GoogleChartsDashboardComponent,
        GoogleChartsControlComponent] });
Ng2GoogleChartsModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.14", ngImport: i0, type: Ng2GoogleChartsModule, providers: [
        GoogleChartsLoaderService
    ] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.14", ngImport: i0, type: Ng2GoogleChartsModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [
                        GoogleChartComponent,
                        GoogleChartsDashboardComponent,
                        GoogleChartsControlComponent,
                    ],
                    providers: [
                        GoogleChartsLoaderService
                    ],
                    exports: [
                        GoogleChartComponent,
                        GoogleChartsDashboardComponent,
                        GoogleChartsControlComponent,
                    ]
                }]
        }] });
