// @ks89/angular-modal-gallery/esm2020/lib/components/modal-gallery/modal-gallery-ref.mjs
/*
 The MIT License (MIT)

 Copyright (C) 2017-2021 Stefano Cappa (Ks89)

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights
 to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 copies of the Software, and to permit persons to whom the Software is
 furnished to do so, subject to the following conditions:

 The above copyright notice and this permission notice shall be included in all
 copies or substantial portions of the Software.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NON INFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 SOFTWARE.
 */
import { Subject } from 'rxjs';
/**
 * Class that represents the modal dialog instance.
 * It is returned by the open method.
 */
export class ModalGalleryRef {
    constructor(overlayRef) {
        this.overlayRef = overlayRef;
        this.close = new Subject();
        this.close$ = this.close.asObservable();
        this.show = new Subject();
        this.show$ = this.show.asObservable();
        this.firstImage = new Subject();
        this.firstImage$ = this.firstImage.asObservable();
        this.lastImage = new Subject();
        this.lastImage$ = this.lastImage.asObservable();
        this.hasData = new Subject();
        this.hasData$ = this.hasData.asObservable();
        this.buttonBeforeHook = new Subject();
        this.buttonBeforeHook$ = this.buttonBeforeHook.asObservable();
        this.buttonAfterHook = new Subject();
        this.buttonAfterHook$ = this.buttonAfterHook.asObservable();
    }
    /**
     * Close modal dialog, disposing the Overlay.
     */
    closeModal() {
        this.overlayRef.dispose();
    }
    /**
     * Method to emit close event.
     * @param event ImageModalEvent event payload
     */
    emitClose(event) {
        this.close.next(event);
    }
    /**
     * Method to emit show event.
     * @param event ImageModalEvent event payload
     */
    emitShow(event) {
        this.show.next(event);
    }
    /**
     * Method to emit firstImage event.
     * @param event ImageModalEvent event payload
     */
    emitFirstImage(event) {
        this.firstImage.next(event);
    }
    /**
     * Method to emit lastImage event.
     * @param event ImageModalEvent event payload
     */
    emitLastImage(event) {
        this.lastImage.next(event);
    }
    /**
     * Method to emit hasData event.
     * @param event ImageModalEvent event payload
     */
    emitHasData(event) {
        this.hasData.next(event);
    }
    /**
     * Method to emit buttonBeforeHook event.
     * @param event ImageModalEvent event payload
     */
    emitButtonBeforeHook(event) {
        this.buttonBeforeHook.next(event);
    }
    /**
     * Method to emit buttonAfterHook event.
     * @param event ImageModalEvent event payload
     */
    emitButtonAfterHook(event) {
        this.buttonAfterHook.next(event);
    }
}
